### DISCUSSION TITLE
Time to Battery Full

### DISCUSSION THREAD

**Discussion by Faqterson** [Danie10/gadgeteerza-homeassistant#5] [Ideas]
I am trying to figure out, if it will be possible to calculate the time it will take for a battery to fully charge.

**Comment by Danie10** [comment]
Yes it would be, using basically the same calculation I did, but instead of calculating from current SOC on current energy used, you'd calculate to 100% instead of down to Min SOC set (or empty). Actually neither will be completely accurate as I've discovered if you are not doing a full 100% charge, the cells get out of balance a bit, and when you see 25% it may actually be at 10%. Same for charging, if the cells are not balanced, you'll see it hit 100% and still be charging the battery for a while longer.

I need to actually look at a counter with a warning after 7 days since the last full charge. The battery has a register for it, but I see it is not returning any data. So I may set a counter to just keep running after 100% charge, and warn if that counter reaches 7 whole days.


Sent with Proton Mail secure email.

------- Original Message -------

**Comment by Faqterson** [comment]
I have done the following:

# Time to Full Charge calc    
    time_to_full_battery:
      friendly_name: "Time to full Battery"
      unit_of_measurement: "Hours"
      icon_template: mdi:sort-clock-ascending-outline
      value_template: >-
          {% set battery_capacity = 234 %}
          {% set battery_soc = states('sensor.battery_soc')|float %}
          {% set battery_current = states('sensor.victron_system_battery_current')|float %}
          {% if (battery_soc) > 99 %}
            {{ 0 }}
          {% elif (battery_current) < 0 %}
            {{ 0 }}
          {% else %}
            {{ (((battery_capacity)-(battery_capacity*battery_soc/100))/battery_current | float ) | round(2) }}
          {% endif %}

I aware the accuracy will be a problem with the battery. I mostly just looking at this as a guide line. 

I have broken time_left_on_battery_min_soc as well by using the battery current and if the battery not doing anything and sitting at 0 A. This breaks the calculation. Will have to add a second if statement to check battery current. 
Using the ac_load was causing problems as the ac_load doesn't always have the same value being used by the battery. 
sensor.victron_system_battery_power is more accurate, just had problem with the negative values.

**Reply by Danie10** [reply]
Good going! OK I think I had the same issue with the 0 A current. Suspect I did do some work around for that in my code. I'm going to have a good look through the registers again today, as I want to put a widget on for time since last full charge, which also needs to look at battery current coming close to 0 A.

**Reply by Danie10** [reply]
Based on my testing today on trying to determine when the battery is actually properly full and balanced (see discussion detail at https://github.com/Danie10/gadgeteerza-homeassistant/discussions/6), I realised that SOC at 100% is not full. It charges another good 40 mins or so before it reaches battery idle for 9 secs. So I'm thinking the estimation of time to reach proper full is not that easy. Reaching 100% is not actually full. It spends its time then on a slower absorption mode, as it slowly drops the charge rate down and cycles between gentle charge & discharge, idle cycles.

The more cell imbalance there is, the longer this 100% stage can last.

I'm going to keep thinking a bit about this thread, on what may be the best thing to project against, so it is meaningful.

Battery voltage is another option to measure to "full" but it is not a linear relationship. It is more logarithmic of nature with the last part being very low current, much longer time.

**Comment by Faqterson** [comment]
Here are some of the changes I made:

    # Calculate hours left to empty based on battery_max_capacity at 100% charge
    # UPDATE MAY 2023: Empty is actually 11% SOC so adjusted formula to reach 11% instead of 0%
    # Hours left = (SOC%-11) * battery_max_capacity (Ah) divided by Current in Amps eg. 86% at 11 Amps is (0.86*206)/11
    # Condition 1 to check battery is discharging (current in Amps will be negative number), then use ABS of battery_current to make negative current flow, a positive float number. Set this to -1 Amp (50W at 52V) instead of 0 Amps to prevent very small discharges when full, from projecting extremely long hours on gauge
    # Condition 2 to check if battery current is zero or charging, then calculate time left based on ac load (AC Watts converted to DC Amps current - because AC Watts = DC Watts)
    # Had to convert all to floats for precision of partial hours, then rounded to 2 decimals
    time_left_on_battery_zero:
      friendly_name: "Time left on Battery to Empty"
      unit_of_measurement: "Hours"
      icon_template: mdi:sort-clock-ascending-outline
      value_template: >-
          {% set battery_current = (states('sensor.battery_current')|float)*-1 %}
          {% set ac_current = (states('sensor.victron_system_bus_charge_current')|float)*-1 %}
          {% set battery_soc = states('sensor.battery_soc')|float %}
          {% set loads_current = states('sensor.ac_loads')|float/states('sensor.battery_voltage')|float %}
          {% set battery_max_capacity = states('sensor.solar_battery_maximum_capacity')|float %}
          {% if (battery_current) <= 1.5 %}
            {{ ((((battery_soc-20)*battery_max_capacity)/100)/ac_current|float ) | round(2) }}
          {% else %}
            {{ ((((battery_soc-20)*battery_max_capacity)/100)/battery_current|float ) | round(2) }}
          {% endif %}
          
    # Calculate hours left to current min SOC based on battery_max_capacity at 100% charge, ie. when it reverts back to grid power
    # Hours left = (SOC%-MinSOC) * battery_max_capacity (Ah) divided by Current in Amps eg. 86%-40% at 11 Amps is (0.46*206)/11
    # Condition 1 to check if SoC has not already reached MinSOC or below it, as then time to reach MinSOC has reached 0 hours, so make 0
    # Condition 2 to check battery is discharging (current in Amps will be negative number), then use ABS of battery_current to make negative current flow, a positive float number.  Set this to -1 Amp (50W at 52V) instead of 0 Amps to prevent very small discharges when full, from projecting extremely long hours on gauge
    # Condition 3 to check if battery current is zero or charging, then calculate time left based on ac load (AC Watts converted to DC Amps current - because AC Watts = DC Watts) 
    # Had to convert all to floats for precision of partial hours, then rounded to 2 decimals
    time_left_on_battery_min_soc:
      friendly_name: "Time left on Battery to Min SoC"
      unit_of_measurement: "Hours"
      icon_template: mdi:sort-clock-ascending
      value_template: >-
        {% set battery_current = (states('sensor.battery_current')|float)*-1 %}
        {% set ac_current = (states('sensor.victron_system_bus_charge_current')|float)*-1 %}
        {% set battery_soc = states('sensor.battery_soc')|float %}
        {% set loads_current = states('sensor.victron_system_battery_power')|float/states('sensor.battery_voltage')|float %}
        {% set battery_min_soc = states('sensor.victron_battery_min_soc')|float %}
        {% set battery_max_capacity = states('sensor.solar_battery_maximum_capacity')|float %}
        {% if (battery_soc-battery_min_soc) < 0 %}
          {{ 0 }}
        {% elif (battery_current) <= 1.5 %}
          {{ ((((battery_soc-battery_min_soc)*battery_max_capacity)/100)/ac_current | float ) | round(2) }}
        {% else %}
          {{ ((((battery_soc-battery_min_soc)*battery_max_capacity)/100)/battery_current | float ) | round(2) }}
        {% endif %}
    
    # Time to Full Charge calc    
    time_to_full_battery:
      friendly_name: "Time to full Battery"
      unit_of_measurement: "Hours"
      icon_template: mdi:sort-clock-ascending-outline
      value_template: >-
          {% set battery_capacity = 234 %}
          {% set battery_soc = states('sensor.battery_soc')|float %}
          {% set battery_current = states('sensor.victron_system_battery_current')|float %}
          {% if (battery_soc) > 99 %}
            {{ 0 }}
          {% elif (battery_current) <= 0 %}
            {{ 0 }}
          {% else %}
            {{ (((battery_capacity)-(battery_capacity*battery_soc/100))/battery_current | float ) | round(2) }}
          {% endif %}

type: horizontal-stack
cards:
  - type: gauge
    entity: sensor.time_to_full_battery
    name: Time to Battery Full
    needle: true
    severity:
      green: 0
      yellow: 8
      red: 12
    max: 15
  - type: gauge
    entity: sensor.victron_battery_soc
    needle: true
    name: Battery Storage
    severity:
      green: 50
      yellow: 20
      red: 0

Having the exact battery time to full charge will be almost impossible because of the BMS and battery balancer. Secondly the Solar output keeps changing during the day, it will also be affected by the house consumption.  
I am happy to have it as a guide line. 

Would love to have the decimal values in Hour:Min but I can't get it working. Keeps complaining about it looing for a float or Integer value in the logs.

**Reply by Danie10** [reply]
Yes the best is to work off battery current as that directly relates to the charge time. The only problem with it, is it gets affected by AC loads that will reduce it if solar output does not exceed battery charge + current. With the time to go to empty, it is easier as the AC load remains a constant (or one variable I should say) draw off the system (so I'm actually calculating estimated time to empty based on solar/grid both being off). With charging, it is more difficult, as the input to battery depends on the balance between grid/solar input as well as AC loads = it varies a lot more.

My installer also gave me a rule of thumb yesterday on charging time, saying that once the battery just hits 100% SOC, I can assume another one hour of charging time to ensure a proper full charge and balanced cells.

BTW there is a register that I'm using to draw the battery capacity - I had a sensor called sensor.solar_battery_maximum_capacity that references it. But not all systems are giving register data equally.

So for float etc errors, I normally spend a lot of time in the Developer tools and States view. Bit by bit, I build each part of the calculation up. For a register variable, thought to be used by name, it has sometimes meant rebooting HA. But yes that would be a type mismatch somewhere.

Another thought may be to get averages of the battery current over a time, and then use that average to calculate. The average would be a moving one, so if you said average per minute, it would update only every minute based on the last minute's average current.

So in summary, two challenges I see to meaningfully calculating an estimated time to full charge are:
1. The battery charge current varying according to AC loads changes = maybe calculating an average of the last minute or two of battery current, and using that as it will be more stable.
2. 100% SOC is actually only the start of the Absorption phase = maybe add 45 mins onto the estimated charge time to 100% SOC, as the real estimated full charge time.
